<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Analyzing binned ChIP-seq/Cut&amp;Run data with ChIPbinner</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Analyzing binned ChIP-seq/Cut&amp;Run data
with ChIPbinner</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(ChIPbinner)</span></code></pre></div>
<div id="database-for-chipbinner" class="section level1">
<h1>Database for ChIPbinner</h1>
<p>Only toy datasets, which have been largely downsampled, are included
with the package. For complete example datasets please see
<a href="https://github.com/padilr1/ChIPbinner_database.git">ChIPbinner
database</a>.</p>
</div>
<div id="example-workflow-for-generating-binned-bed-files" class="section level1">
<h1>Example workflow for generating binned BED files</h1>
<p>For input into ChIPbinner, the sequenced samples will need to be
binned into windows, and formatted as a
<a href="https://genome.ucsc.edu/FAQ/FAQformat">BED</a> file. One of the
tools that can be used to perform this task is
<a href="https://bedtools.readthedocs.io/en/latest/">BEDtools</a>, which
can convert an aligned sequence file, typically found in a
<a href="https://genome.ucsc.edu/goldenPath/help/bam.html">BAM</a>
format, into a binned BED file as demonstrated below.</p>
<p>Reference windows ($WINDOW_REF) can be found in the
<a href="https://github.com/padilr1/ChIPbinner_database.git">ChIPbinner
database</a>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># convert aligned BAM file to BED file</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">bedtools</span> bamtobed <span class="at">-i</span> <span class="va">${sample}</span>.sorted.bam <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-k1,1</span> <span class="at">-k2,2n</span> <span class="op">&gt;</span> <span class="va">${sample}</span>.bed</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co"># intersect sample BED file with binned reference files</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-a</span> <span class="va">$WINDOW_REF</span> <span class="at">-b</span> <span class="va">${sample}</span>.bed <span class="at">-sorted</span> <span class="at">-c</span> <span class="at">-nonamecheck</span> <span class="op">&gt;</span> <span class="va">$OUTPUT_NAME</span>.binned.bed</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="co"># example for binning Cal27 WT sample into 10kb windows</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="ex">bedtools</span> bamtobed <span class="at">-i</span> Cal27.WT.H3K36me2.sorted.bam <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-k1,1</span> <span class="at">-k2,2n</span> <span class="op">&gt;</span> Cal27.WT.H3K36me2.bed</span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="ex">bedtools</span> intersect <span class="at">-a</span> hg38.10kb.windows.bed <span class="at">-b</span> Cal27.WT.H3K36me2.bed <span class="at">-sorted</span> <span class="at">-c</span> <span class="at">-nonamecheck</span> <span class="op">&gt;</span> Cal27.WT.H3K36me2.10kb.bed</span></code></pre></div>
</div>
<div id="generating-normalized-bigwig-files" class="section level1">
<h1>Generating normalized bigwig files</h1>
<p>When you bin your files into windows, especially from aligned (BAM or
SAM) files, they are typically raw read counts. They will need to be
normalized by the input (highly recommended for ChIP-seq data) and
likely scaled by some factor (we recommend using a normalization scaling
factor like genome-wide modification percentage information obtained
from mass spectrometry, Drosophila spike-in or any other types of
spike-in). This step will have to be performed prior to moving on to the
rest of the workflow.</p>
<p>Here, we are generating normalized bigwig files for 10kb-binned files
(in BED format) for a head and neck squamous cell carcinoma (HNSCC) line
(Cal27) for the broad histone mark H3K36me2 generated via ChIP-seq. The
lysine 36 methyl transferase NSD1 was knocked out in the sample
(NSD1-KO) and is being compared to the wildtype (WT) cell line. Please
see
<a href="https://www.sciencedirect.com/science/article/pii/S2211124721000826?via%3Dihub">Farhangdoost
et al. 2021</a> for more details.</p>
<p><b>Important note:</b> For ChIPbinner, only two genome assemblies can
be used: <code>hg38</code> or <code>mm10</code>. We will be using the
<code>hg38</code> assembly for this analysis.</p>
<p>If you’ve ran the code chunk successfully, your resulting bigwig
files will be found in the <code>output_directory</code> that you
specified when running the function. The
<code>immunoprecipitated_binned_file</code> refers to the
immunoprecipitated sample for ChIP-seq or the targeted, enriched sample
for CUT&amp;RUN/TAG. The <code>input_binned_file</code> refers to the
genomic input for ChIP-seq samples or the IgG control sample for
CUT&amp;RUN/TAG.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># generate normalized bigwig for WT sample</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">norm_bw</span>(<span class="at">out_dir =</span> <span class="st">&quot;output_directory&quot;</span>, <span class="at">genome_assembly =</span> <span class="st">&quot;hg38&quot;</span>,</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>    <span class="at">immunoprecipitated_binned_file =</span> <span class="st">&quot;Cal27.WT.H3K36me2.10kb.bed&quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>    <span class="at">use_input =</span> <span class="cn">TRUE</span>, <span class="at">input_binned_file =</span> <span class="st">&quot;Cal27.WT_input.H3K36me2.10kb.bed&quot;</span>,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>    <span class="at">raw_count_cutoff =</span> <span class="dv">0</span>, <span class="at">pseudocount =</span> <span class="fl">1e-15</span>, <span class="at">scaling_factor =</span> <span class="fl">0.450328805</span>)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co"># generate normalized bigwig for NSD1-KO sample</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="fu">norm_bw</span>(<span class="at">out_dir =</span> <span class="st">&quot;output_directory&quot;</span>, <span class="at">genome_assembly =</span> <span class="st">&quot;hg38&quot;</span>,</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>    <span class="at">immunoprecipitated_binned_file =</span> <span class="st">&quot;Cal27.NSD1_KO.H3K36me2.10kb.bed&quot;</span>,</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>    <span class="at">use_input =</span> <span class="cn">TRUE</span>, <span class="at">input_binned_file =</span> <span class="st">&quot;Cal27.NSD1_KO_input.H3K36me2.10kb.bed&quot;</span>,</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>    <span class="at">raw_count_cutoff =</span> <span class="dv">0</span>, <span class="at">pseudocount =</span> <span class="fl">1e-15</span>, <span class="at">scaling_factor =</span> <span class="fl">0.192272095</span>)</span></code></pre></div>
</div>
<div id="generating-scatterplots-with-bins-stratified-into-genic-and-intergenic-regions" class="section level1">
<h1>Generating scatterplots with bins stratified into genic and
intergenic regions</h1>
<p>After normalizing and converting the binned BED files into bigwigs,
we can read them in and generate a scatterplot comparing two cell lines.
The scatterplot will annotate genic and intergenic regions accordingly.
The resulting figures will be found in <code>/output_directory</code>
for the code chunk below.</p>
<p>Genic and intergenic regions for <code>hg38</code> and
<code>mm10</code> are included with the package and can be accessed as
indicated in the code chunk below. <b> Note: you will need to adjust the
minimum and maximum values for the x- and y-axis accordingly. To find
optimal values for the x- and y-axis, run <code>pre_clust</code> to
create an exploratory scatterplot of the samples you’re
comparing.</b></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">genic_intergenic_scatterplot</span>(<span class="at">out_dir =</span> <span class="st">&quot;output_directory&quot;</span>, <span class="at">genome_assembly =</span> <span class="st">&quot;hg38&quot;</span>,</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>    <span class="at">cell_line =</span> <span class="st">&quot;Cal27&quot;</span>, <span class="at">wildtype_samp_norm_bw =</span> <span class="st">&quot;Cal27.WT.H3K36me2.10kb.norm.bw&quot;</span>,</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>    <span class="at">treated_samp_norm_bw =</span> <span class="st">&quot;Cal27.NSD1_KO.H3K36me2.10kb.norm.bw&quot;</span>,</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>    <span class="at">are_R_objects =</span> <span class="cn">FALSE</span>, <span class="at">histone_mark =</span> <span class="st">&quot;H3K36me2&quot;</span>, <span class="at">output_filename =</span> <span class="st">&quot;Cal27.WT_NSD1KO.H3K36me2.10kb&quot;</span>,</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>    <span class="at">title_of_plot =</span> <span class="st">&quot;Cal27 ChIP-seq H3K36me2&quot;</span>, <span class="at">xaxis_label =</span> <span class="st">&quot;WT&quot;</span>,</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>    <span class="at">yaxis_label =</span> <span class="st">&quot;NSD1_KO&quot;</span>, <span class="at">max_x =</span> <span class="dv">1</span>, <span class="at">max_y =</span> <span class="dv">1</span>, <span class="at">min_x =</span> <span class="sc">-</span><span class="dv">5</span>,</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>    <span class="at">min_y =</span> <span class="sc">-</span><span class="dv">5</span>, <span class="at">pow =</span> <span class="fl">1.25</span>, <span class="at">show_scales =</span> <span class="cn">FALSE</span>, <span class="at">show_legend =</span> <span class="cn">TRUE</span>,</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>    <span class="at">legend_pos =</span> <span class="st">&quot;left&quot;</span>)</span></code></pre></div>
</div>
<div id="exploratory-analyses" class="section level1">
<h1>Exploratory analyses</h1>
<div id="pca" class="section level2">
<h2>PCA</h2>
<p>Plots of principal component analysis can be generated using
<code>plot_PCA</code> with an example below. PCA plots are useful for
assessing the treatment effect and the consistency of replicates.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>  <span class="fu">plot_PCA</span>(</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>    <span class="at">out_dir =</span> <span class="st">&quot;outdir&quot;</span>,</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>    <span class="at">output_filename =</span> <span class="st">&quot;PCA_plot&quot;</span>,</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>    <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;blue3&quot;</span>,<span class="st">&quot;red&quot;</span>,<span class="st">&quot;red4&quot;</span>,<span class="st">&quot;forestgreen&quot;</span>,<span class="st">&quot;green1&quot;</span>),</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>    <span class="at">sample_labels =</span> <span class="fu">c</span>(<span class="st">&quot;samp_1&quot;</span>, <span class="st">&quot;samp_2&quot;</span>, <span class="st">&quot;samp_3&quot;</span>, <span class="st">&quot;samp_4&quot;</span>, <span class="st">&quot;samp_5&quot;</span>, <span class="st">&quot;samp_6&quot;</span>),</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>    <span class="at">plot_title =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>    <span class="at">plot_height =</span> <span class="dv">8</span>,</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>    <span class="at">plot_width=</span> <span class="dv">12</span>,</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>    <span class="st">&quot;samp1_10kb.norm.bw&quot;</span>,</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>    <span class="st">&quot;samp2_10kbb.norm.bw&quot;</span>,</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>    <span class="st">&quot;samp3_10kbb.norm.bw&quot;</span>,</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>    <span class="st">&quot;samp4_10kbb.norm.bw&quot;</span>,</span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>    <span class="st">&quot;samp5_10kbb.norm.bw&quot;</span>,</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a>    <span class="st">&quot;samp6_10kbb.norm.bw&quot;</span>,</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>  )</span></code></pre></div>
</div>
<div id="correlation-plots" class="section level2">
<h2>Correlation plots</h2>
<p>Correlation plots can be generated using
<code>plot_correlation</code> with an example below. Correlation plots
are useful for assessing separation of samples according to treatment
and consistency of replicates.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">plot_correlation</span>(</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>    <span class="at">out_dir =</span> <span class="st">&quot;outdir&quot;</span>,</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>    <span class="at">output_filename =</span> <span class="st">&quot;correlation_plot&quot;</span>,</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>    <span class="at">sample_labels =</span> <span class="fu">c</span>(<span class="st">&quot;samp_1&quot;</span>, <span class="st">&quot;samp_2&quot;</span>, <span class="st">&quot;samp_3&quot;</span>, <span class="st">&quot;samp_4&quot;</span>, <span class="st">&quot;samp_5&quot;</span>, <span class="st">&quot;samp_6&quot;</span>),</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>    <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;blue3&quot;</span>,<span class="st">&quot;red&quot;</span>,<span class="st">&quot;red4&quot;</span>,<span class="st">&quot;forestgreen&quot;</span>,<span class="st">&quot;green1&quot;</span>),</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>    <span class="at">plot_height =</span> <span class="dv">4</span>,</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>    <span class="at">plot_width =</span> <span class="dv">4</span>,</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>    <span class="st">&quot;samp1_10kb.norm.bw&quot;</span>,</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>    <span class="st">&quot;samp2_10kbb.norm.bw&quot;</span>,</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>    <span class="st">&quot;samp3_10kbb.norm.bw&quot;</span>,</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>    <span class="st">&quot;samp4_10kbb.norm.bw&quot;</span>,</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>    <span class="st">&quot;samp5_10kbb.norm.bw&quot;</span>,</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>    <span class="st">&quot;samp6_10kbb.norm.bw&quot;</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>  )</span></code></pre></div>
</div>
</div>
<div id="combining-replicates" class="section level1">
<h1>Combining replicates</h1>
<p>Replicates for a given treatment can be combined using
<code>merge_norm_bw</code>. The merged bigWig file can be used for
downstream analysis.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">merge_norm_bw</span>(<span class="at">rep1 =</span> <span class="st">&quot;WT_rep1_10kb_norm.bw&quot;</span>,</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>              <span class="at">rep2 =</span> <span class="st">&quot;WT_rep2_10kb_norm.bw&quot;</span>,</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>              <span class="at">merged_samples_label =</span> <span class="st">&quot;merged_WT&quot;</span>,</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>              <span class="at">out_dir =</span> <span class="st">&quot;outdir&quot;</span>)</span></code></pre></div>
</div>
<div id="identifying-clusters-of-similarly-behaving-genomic-bins" class="section level1">
<h1>Identifying clusters of similarly-behaving genomic bins</h1>
<div id="processing-bigwig-files-for-clustering" class="section level2">
<h2>Processing bigwig files for clustering</h2>
<p>Prior to generating clusters, we need to pre-process the normalized
bigWig files, which entails finding overlapping regions between two
bigWig files, removing the bottom and top 1% of bins across the two
samples (trimming to make the graphics more visible) and finally
generating a matrix of scores as well as a BED file of coordinates to be
used with HDBScan for generating clusters of similarly-behaving bins.
This is accomplished using the <code>pre_clus()</code> function. We will
continue to use the samples from the previous steps above.</p>
<p><b>Important note: Normally, bigWig files are used as input
throughout the workflow. However, R objects of the bigWig files can also
be used, but must specified by setting
<code>are_R_objects=TRUE</code></b>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="fu">pre_clust</span>(</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>  <span class="at">out_dir =</span> <span class="st">&quot;output_directory&quot;</span>,</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  <span class="at">treated_samp_norm_bw =</span> <span class="st">&quot;Cal27.WT.H3K36me2.10kb.norm.bw&quot;</span>,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="at">wildtype_samp_norm_bw =</span> <span class="st">&quot;Cal27.NSD1_KO.H3K36me2.10kb.norm.bw&quot;</span>,</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>  <span class="at">output_filename =</span> <span class="st">&quot;Cal27.WT.NSD1_KO.H3K36me2.10kb&quot;</span>,</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>  <span class="at">are_R_objects =</span> <span class="cn">FALSE</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="running-hdbscan-to-generate-clusters" class="section level2">
<h2>Running HDBSCAN to generate clusters</h2>
<p>The <code>clus()</code> function can be used to generate clusters of
similarly-behaving genomic bins. It is based on an unsupervised learning
algorithm to find clusters, or dense regions, of a dataset. More details
on the Hierarchical Density-Based Spatial Clustering of Applications
with Noise (HDBSCAN) algorithm can be found
<a href="https://hdbscan.readthedocs.io/en/latest/index.html">here</a>.
The function uses <code>reticulate</code> as an interface to a Python
session. <b>Python version 3.9 will need to be installed in your
environment prior to running the function.</b> No other actions are
required from the user, as the minimal Python packages are automatically
installed in an isolated virtual environment specifically designated for
the ChIPbinner R package. This minimizes the risk of inadvertently
disrupting another Python installation on the user’s system.</p>
<p><b>It is important to consider this step will require substantial
processing memory as well as time </b>. Thus, it is recommended that you
use the python-implementation of HDBSCAN in a High-performance Computing
server instead. Nevertheless, an example workflow for running HDBSCAN is
found below. Here, we use the complete matrix file, which required
substantial processing memory and multiple parallel cores, typically
found in a high-performance computing cluster. The complete matrix file
can be found at the
<a href="https://github.com/padilr1/ChIPbinner_database.git">ChIPbinner
database</a>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># using the complete matrix</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">generate_clust</span>(</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="at">output_file_name =</span> <span class="st">&quot;Cal27.WT.NSD1_KO.H3K36me2.10kb&quot;</span>,</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="at">out_dir =</span> <span class="st">&quot;output_directory&quot;</span>,</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>  <span class="at">matrix_file =</span> <span class="st">&quot;Cal27.WT.NSD1_KO.H3K36me2.10kb.mat.csv&quot;</span>,</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>  <span class="at">minpts =</span> <span class="dv">5000</span>,</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>  <span class="at">minsamps =</span> <span class="dv">5000</span>,</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>  <span class="at">cores =</span> <span class="dv">6</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="annotating-clusters" class="section level2">
<h2>Annotating clusters</h2>
<p>After identifying clusters using HDBSCAN, these clusters need to be
annotated for use in downstream analysis.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">annotate_clust</span>(<span class="at">number_of_clusters =</span> <span class="dv">3</span>,</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>               <span class="at">matrix_file =</span> <span class="st">&quot;Cal27.WT.NSD1_KO.H3K36me2.10kb_mat.csv&quot;</span>,</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>               <span class="at">pooled_bed_file =</span> <span class="st">&quot;Cal27.WT.NSD1_KO.H3K36me2.10kb_pooled.bed&quot;</span>,</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>               <span class="at">hdbscan_output_file =</span> <span class="st">&quot;clus.Cal27.WT.NSD1_KO.H3K36me2.10kb.5000.5000.txt&quot;</span>,</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>               <span class="at">output_filename =</span> <span class="st">&quot;Cal27.WT.NSD1_KO.H3K36me2.10kb&quot;</span>,</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>               <span class="at">out_dir=</span><span class="st">&quot;output_directory&quot;</span>)</span></code></pre></div>
</div>
<div id="generating-density-based-scatterplots" class="section level2">
<h2>Generating density-based scatterplots</h2>
<p>Finally, a density-based cluster scatterplot (along with the
genic/intergenic scatterplot) can be generated using
<code>density_based_scatterplot()</code>. Here, we used the complete
matrix of enrichment scores to generate the HBDSCAN output. The complete
HDBSCAN output can found at
<a href="https://github.com/padilr1/ChIPbinner_database.git">ChIPbinner
database</a>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">density_based_scatterplot</span>(</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>  <span class="at">out_dir =</span> <span class="st">&quot;output_directory&quot;</span>,</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>  <span class="at">genome_assembly =</span> <span class="st">&quot;hg38&quot;</span>,</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>  <span class="at">are_R_objects =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  <span class="at">output_filename =</span> <span class="st">&quot;Cal27.WT.NSD1_KO.H3K36me2.10kb&quot;</span>,</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>  <span class="at">wildtype_samp_norm_bw =</span><span class="st">&quot;Cal27_WT_H3K36me2_10kb.norm.bw&quot;</span>,</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>  <span class="at">treated_samp_norm_bw =</span> <span class="st">&quot;Cal27_NSD1_KO_H3K36me2_10kb.norm.bw&quot;</span>,</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>  <span class="at">cell_line =</span> <span class="st">&quot;Cal27&quot;</span>,</span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>  <span class="at">histone_mark =</span> <span class="st">&quot;H3K36me2&quot;</span>,</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>  <span class="at">annotated_clusters =</span><span class="st">&quot;Cal27.WT.NSD1_KO.H3K36me2.10kb_annotated_clusters.rda&quot;</span>,</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>  <span class="at">number_of_clusters =</span> <span class="dv">3</span>,</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>  <span class="at">title_of_plot =</span> <span class="st">&quot;H3K36me2&quot;</span>,</span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>  <span class="at">pow =</span> <span class="fl">1.1</span>,</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a>  <span class="at">show_legend =</span> <span class="cn">TRUE</span>,</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a>  <span class="at">min =</span> <span class="sc">-</span><span class="dv">5</span>,</span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a>  <span class="at">max =</span> <span class="dv">2</span>,</span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a>  <span class="at">bin_size =</span> <span class="dv">50</span>,</span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>  <span class="at">show_scales =</span> <span class="cn">FALSE</span>,</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a>  <span class="at">xaxis_label =</span> <span class="st">&quot;WT&quot;</span>,</span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a>  <span class="at">yaxis_label =</span> <span class="st">&quot;NSD1_KO&quot;</span>,</span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a>  <span class="at">height_of_figure =</span> <span class="dv">6</span>,</span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a>  <span class="at">width_of_figure =</span> <span class="dv">15</span></span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="overlap-enrichmentdepletion-analysis-with-annotated-clusters-of-bins" class="section level1">
<h1>Overlap enrichment/depletion analysis with annotated clusters of
bins</h1>
<p>After identifying and annotating clusters of bins, a Fisher’s exact
test can be used to determine whether these bins overlap a specific
class of annotated regions against a background of all bins. The
underlying algorithm is based on
<a href="https://bioconductor.org/packages/release/bioc/html/LOLA.html">LOLA
- locus overlap analysis for enrichment of genomic ranges</a>. The
example below evaluates overlap with
<a href="https://useast.ensembl.org/info/genome/index.html">Ensembl</a>
annotations genome-wide. However, the user can also evaluate overlap
with other functional annotations by inputting an R object of the
database of their choice in the parameter
<code>functional_db</code>.</p>
<p>The
<a href="https://github.com/padilr1/ChIPbinner_database.git">ChIPbinner
database</a> contains curated databases ready for input into the
parameter <code>functional_db</code> within the
<code>enrich_clust()</code> function.</p>
<p>Nevertheless, by specifying the <code>region</code> to be either
<code>genic</code> or <code>intergenic</code>, the user can evaluate
exclusively genic or intergenic bins overlapping a specific class of
annotated regions. In these cases, the background is stratified to only
genic or intergenic regions to avoid spurious associations to
annotations confounded by their predominantly genic or intergenic
localization.</p>
<p>Below is an example of how to calculate overlap enrichment &amp;
depletion result of Ensembl annotations with bins found in cluster B
identified from the previous analyses. Note that the size of the dots
corresponds to the number of bins overlapping the corresponding
annotation. The p-value is based on Fisher’s exact test of bins
overlapping a specific class of annotated regions versus a background of
all bins found in both cell lines being compared.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">enrich_clust</span>(</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>  <span class="at">genome_assembly =</span> <span class="st">&quot;hg38&quot;</span>,</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>  <span class="at">annotated_clusters =</span> <span class="st">&quot;Cal27.WT.NSD1_KO.H3K36me2.10kb_annotated_clusters.rda&quot;</span>,</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>  <span class="at">query_cluster =</span> <span class="st">&quot;B&quot;</span>,</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>  <span class="at">pooled_bed_file =</span> <span class="st">&quot;Cal27.WT.NSD1_KO.H3K36me2.10kb_pooled.bed&quot;</span>,</span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>  <span class="at">functional_db =</span> <span class="st">&quot;hg38_ensemblDB.rda&quot;</span>,</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>  <span class="at">region =</span> <span class="st">&quot;genome_wide&quot;</span>,</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a>  <span class="at">cores =</span> <span class="dv">1</span>,</span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a>  <span class="at">n_elements =</span> <span class="dv">7</span>,</span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a>  <span class="at">cutoff_for_overlap =</span> <span class="dv">1000</span>,</span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a>  <span class="at">file_plot_name =</span> <span class="st">&quot;ensembl.genome_wide.enrichment_depletion&quot;</span>,</span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a>  <span class="at">output_table_name =</span> <span class="st">&quot;ensembl.genome_wide.enrichment_depletion&quot;</span>,</span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a>  <span class="at">width_of_plot =</span> <span class="dv">7</span>,</span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a>  <span class="at">height_of_plot =</span> <span class="dv">3</span>,</span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a>  <span class="at">out_dir =</span> <span class="st">&quot;output_directory&quot;</span></span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a>  )</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
