<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate density-based clusters. — density_based_scatterplot • ChIPbinner</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Generate density-based clusters. — density_based_scatterplot"><meta property="og:description" content="This function identifies similarly-behaving genomic compartments, such as regions with the highest loss or gain of a specific histone mark."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ChIPbinner</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ChIPbinner_workflow.html">Analyzing binned ChIP-seq/Cut&amp;Run data with ChIPbinner</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate density-based clusters.</h1>
    
    <div class="hidden name"><code>density_based_scatterplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function identifies similarly-behaving genomic compartments, such as regions with the highest loss or gain of a specific histone mark.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">density_based_scatterplot</span><span class="op">(</span></span>
<span>  <span class="va">out_dir</span>,</span>
<span>  <span class="va">genome_assembly</span>,</span>
<span>  <span class="va">treated_samp_norm_bw</span>,</span>
<span>  <span class="va">wildtype_samp_norm_bw</span>,</span>
<span>  are_R_objects <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">cell_line</span>,</span>
<span>  <span class="va">histone_mark</span>,</span>
<span>  <span class="va">annotated_clusters</span>,</span>
<span>  <span class="va">number_of_clusters</span>,</span>
<span>  <span class="va">output_filename</span>,</span>
<span>  <span class="va">title_of_plot</span>,</span>
<span>  plot_title_font_size <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  pow <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  show_legend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  legend_font_size <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  <span class="va">min</span>,</span>
<span>  <span class="va">max</span>,</span>
<span>  <span class="va">bin_size</span>,</span>
<span>  show_scales <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">xaxis_label</span>,</span>
<span>  <span class="va">yaxis_label</span>,</span>
<span>  axis_title_font_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  height_of_figure <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  width_of_figure <span class="op">=</span> <span class="fl">15</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>out_dir</dt>
<dd><p>Output directory for the scatterplot.</p></dd>


<dt>genome_assembly</dt>
<dd><p>Must of be one of hg38 or mm10.</p></dd>


<dt>treated_samp_norm_bw</dt>
<dd><p>The normalized bigwig file for the treated sample.</p></dd>


<dt>wildtype_samp_norm_bw</dt>
<dd><p>The normalized bigwig file for the wildtype sample.</p></dd>


<dt>are_R_objects</dt>
<dd><p>Boolean term (true or false) to indicate if the inputted bigwig files are R objects. It'll use load() for the reps as opposed to reading them in via rtracklayer::import.bed(). Default to FALSE.</p></dd>


<dt>cell_line</dt>
<dd><p>The cell line of the samples.</p></dd>


<dt>histone_mark</dt>
<dd><p>The broad histone mark used for the analysis.</p></dd>


<dt>annotated_clusters</dt>
<dd><p>The annotated clusters generated using annotate_clust(), which is an R object that needs to be loaded.</p></dd>


<dt>number_of_clusters</dt>
<dd><p>The total number of clusters identified by the HDBSCAN algorithm.</p></dd>


<dt>output_filename</dt>
<dd><p>Filename for the resuting scatterplot.</p></dd>


<dt>title_of_plot</dt>
<dd><p>The title of the plot.</p></dd>


<dt>plot_title_font_size</dt>
<dd><p>An integer specifying the font size for each plot title.</p></dd>


<dt>pow</dt>
<dd><p>The power of. Returns the value of x to the power of y (x^y) and this is used for the scales (i.e. show bins if surpassing a certain intensity). Defaults to 1.25.</p></dd>


<dt>show_legend</dt>
<dd><p>Boolean term whether to show legend or not.</p></dd>


<dt>legend_font_size</dt>
<dd><p>An integer specifying the font size for the legends.</p></dd>


<dt>min</dt>
<dd><p>The min for the x and y axis.</p></dd>


<dt>max</dt>
<dd><p>The max for the x and y axis</p></dd>


<dt>bin_size</dt>
<dd><p>The number of bins represented by each hexagonal shape.</p></dd>


<dt>show_scales</dt>
<dd><p>Boolean term whether to show scales or not.</p></dd>


<dt>xaxis_label</dt>
<dd><p>Label for the x-axis. This is normally the baseline label (i.e. WT)</p></dd>


<dt>yaxis_label</dt>
<dd><p>Label for the y-axis. This is normally the treated sample label.</p></dd>


<dt>axis_title_font_size</dt>
<dd><p>An integer specifying the font size for the axis titles.</p></dd>


<dt>height_of_figure</dt>
<dd><p>A numeric specifying the height of the plots.</p></dd>


<dt>width_of_figure</dt>
<dd><p>A numeric specifying the width of the plots.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a plot of density-based clusters.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Title</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://lattice.r-forge.r-project.org/" class="external-link">lattice</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.cruk.cam.ac.uk/core-facilities/bioinformatics-core/software/DiffBind" class="external-link">DiffBind</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Linking to GEOS 3.12.0, GDAL 3.8.0, PROJ 9.3.0; sf_use_s2() is TRUE</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.stats.ox.ac.uk/pub/MASS4/" class="external-link">MASS</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘MASS’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:patchwork’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     area</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:dplyr’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     select</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/r-spatial/lwgeom/" class="external-link">lwgeom</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Linking to liblwgeom 3.0.0beta1 r16016, GEOS 3.12.0, PROJ 9.3.1</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>PROJ versions differ: lwgeom has 9.3.1 sf has 9.3.0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘lwgeom’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following object is masked from ‘package:sf’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     st_perimeter</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggrepel.slowkow.com/" class="external-link">ggrepel</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/edzer/hexbin" class="external-link">hexbin</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/VPetukhov/ggrastr" class="external-link">ggrastr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: viridisLite</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://kwstat.github.io/pals/" class="external-link">pals</a></span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘pals’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:viridis’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     cividis, inferno, magma, plasma, turbo, viridis</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:viridisLite’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     cividis, inferno, magma, plasma, turbo, viridis</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://isoband.r-lib.org" class="external-link">isoband</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># using data in /extdata folder</span></span></span>
<span class="r-in"><span><span class="co"># system.file("extdata", "downsampled.Cal27.WT_input.H3K36me2.10kb.bed.gz", package = "ChIPbinner")</span></span></span>
<span class="r-in"><span><span class="fu">density_based_scatterplot</span><span class="op">(</span></span></span>
<span class="r-in"><span>  out_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"ChIPbinner"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  genome_assembly <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>  are_R_objects <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  output_filename <span class="op">=</span> <span class="st">"complete"</span>,</span></span>
<span class="r-in"><span>  wildtype_samp_norm_bw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/data"</span>,<span class="st">"Cal27_WT_H3K36me2_10kb_norm_bw.rda"</span>, package <span class="op">=</span> <span class="st">"ChIPbinner"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  treated_samp_norm_bw <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/data"</span>,<span class="st">"Cal27_NSD1_KO_H3K36me2_10kb_norm_bw.rda"</span>, package <span class="op">=</span> <span class="st">"ChIPbinner"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  cell_line <span class="op">=</span> <span class="st">"Cal27"</span>,</span></span>
<span class="r-in"><span>  histone_mark <span class="op">=</span> <span class="st">"H3K36me2"</span>,</span></span>
<span class="r-in"><span>  annotated_clusters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/data"</span>,<span class="st">"complete_annotated_clusters.rda"</span>, package <span class="op">=</span> <span class="st">"ChIPbinner"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  number_of_clusters <span class="op">=</span> <span class="fl">3</span>,</span></span>
<span class="r-in"><span>  title_of_plot <span class="op">=</span> <span class="st">"H3K36me2"</span>,</span></span>
<span class="r-in"><span>  pow <span class="op">=</span> <span class="fl">1.1</span>,</span></span>
<span class="r-in"><span>  show_legend <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  min <span class="op">=</span> <span class="op">-</span><span class="fl">5</span>,</span></span>
<span class="r-in"><span>  max <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>  bin_size <span class="op">=</span> <span class="fl">50</span>,</span></span>
<span class="r-in"><span>  show_scales <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  xaxis_label <span class="op">=</span> <span class="st">"WT"</span>,</span></span>
<span class="r-in"><span>  yaxis_label <span class="op">=</span> <span class="st">"NSD1_KO"</span>,</span></span>
<span class="r-in"><span>  height_of_figure <span class="op">=</span> <span class="fl">6</span>,</span></span>
<span class="r-in"><span>  width_of_figure <span class="op">=</span> <span class="fl">15</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Density-based scatterplots generated!"</span>
<span class="r-in"><span><span class="co"># using data in /data folder</span></span></span>
<span class="r-in"><span><span class="co"># density_based_scatterplot(</span></span></span>
<span class="r-in"><span><span class="co">#   out_dir = system.file("extdata", package = "ChIPbinner"),</span></span></span>
<span class="r-in"><span><span class="co">#   genome_assembly = "hg38",</span></span></span>
<span class="r-in"><span><span class="co">#   are_R_objects = TRUE,</span></span></span>
<span class="r-in"><span><span class="co">#   output_filename = "complete",</span></span></span>
<span class="r-in"><span><span class="co">#   wildtype_samp_norm_bw = system.file("data/Cal27_WT_H3K36me2_10kb_norm_bw.rda", package = "ChIPbinner"),</span></span></span>
<span class="r-in"><span><span class="co">#   treated_samp_norm_bw = system.file("data/Cal27_NSD1_KO_H3K36me2_10kb_norm_bw.rda", package = "ChIPbinner"),</span></span></span>
<span class="r-in"><span><span class="co">#   cell_line = "Cal27",</span></span></span>
<span class="r-in"><span><span class="co">#   histone_mark = "H3K36me2",</span></span></span>
<span class="r-in"><span><span class="co">#   annotated_clusters = system.file("data/complete_annotated_clusters.rda", package = "ChIPbinner"),</span></span></span>
<span class="r-in"><span><span class="co">#   number_of_clusters = 3,</span></span></span>
<span class="r-in"><span><span class="co">#   title_of_plot = "H3K36me2",</span></span></span>
<span class="r-in"><span><span class="co">#   pow = 1.1,</span></span></span>
<span class="r-in"><span><span class="co">#   show_legend = TRUE,</span></span></span>
<span class="r-in"><span><span class="co">#   min = -5,</span></span></span>
<span class="r-in"><span><span class="co">#   max = 2,</span></span></span>
<span class="r-in"><span><span class="co">#   bin_size = 50,</span></span></span>
<span class="r-in"><span><span class="co">#   show_scales = FALSE,</span></span></span>
<span class="r-in"><span><span class="co">#   xaxis_label = "WT",</span></span></span>
<span class="r-in"><span><span class="co">#   yaxis_label = "NSD1_KO",</span></span></span>
<span class="r-in"><span><span class="co">#   height_of_figure = 6,</span></span></span>
<span class="r-in"><span><span class="co">#   width_of_figure = 15</span></span></span>
<span class="r-in"><span><span class="co"># )</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Reinnier Padilla.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

